test/integration/tcp_proxy_integration_test.cc

```diff
@@ -372,6 +372,341 @@ TEST_P(TcpProxyIntegrationTest, TestIdletimeoutWithLargeOutstandingData) {
   ASSERT_TRUE(fake_upstream_connection->waitForDisconnect(true));
 }
 
+class TcpProxyMetadataMatchIntegrationTest : public TcpProxyIntegrationTest {
+public:
+  void setup(const std::string& tcp_proxy_config, const std::string& endpoint_role,
+             const std::string& endpoint_version, const std::string& endpoint_stage);
+
+  void expectEndpointToMatchRoute();
+  void expectEndpointNotToMatchRoute();
+};
+
+void TcpProxyMetadataMatchIntegrationTest::setup(const std::string& tcp_proxy_config,
+                                                 const std::string& endpoint_role,
+                                                 const std::string& endpoint_version,
+                                                 const std::string& endpoint_stage) {
+
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+    auto* static_resources = bootstrap.mutable_static_resources();
+
+    ASSERT(static_resources->listeners_size() == 1);
+    static_resources->mutable_listeners(0)
+        ->mutable_filter_chains(0)
+        ->mutable_filters(0)
+        ->mutable_typed_config()
+        ->PackFrom(TestUtility::parseYaml<envoy::config::filter::network::tcp_proxy::v2::TcpProxy>(
+            tcp_proxy_config));
+
+    ASSERT(static_resources->clusters_size() == 1);
+    static_resources->clear_clusters();
+    static_resources->add_clusters()->CopyFrom(
+        TestUtility::parseYaml<envoy::api::v2::Cluster>(fmt::format(
+            R"EOF(
+            name: cluster_0
```

Comment:
We prefer programmatic generation of these objects rather than parsing configs (although a number of places still do the latter). Programmatic generation will be easier to update when we move from v2 to v3.

Reply:
working on this

Reply:
done